File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_python/p04028/p04028_num5_parsed.py
MOD = 10**9 + 7

def count_ways(N, s):
    # Initializing the dp table
    dp = [[0] * (N+1) for _ in range(len(s)+1)]

    # Base case
    dp[0][0] = 1

    # Dynamic programming
    for i in range(1, len(s)+1):
        for j in range(N+1):
            dp[i][j] = dp[i-1][j]
            if s[i-1] == '0':
                dp[i][j] += dp[i-1][j-1] if j > 0 else 0
            else:
                dp[i][j] += dp[i-1][j] if j < N else 0
            dp[i][j] %= MOD

    return dp[-1][-1]

# Reading input
N = int(input())
s = input().strip()

# Counting the number of ways
result = count_ways(N, s)

# Printing the result
print(result)

File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_cpp/p04028/p04028_num5_parsed.cpp
#include <iostream>
#include <vector>

const int MOD = 1000000007;

int main() {
    int N;
    std::string s;
    std::cin >> N >> s;

    int M = s.length();

    // Create a table to store the number of ways for each length of string
    std::vector<std::vector<int>> dp(N+1, std::vector<int>(M+1, 0));

    // Initialize the table
    dp[0][0] = 1;

    // Calculate the number of ways for each length of string
    for (int i = 1; i <= N; i++) {
        for (int j = 0; j <= M; j++) {
            if (j > 0) {
                dp[i][j] = dp[i][j-1];
            }
            if (s[j] == '0') {
                dp[i][j] = (dp[i][j] + dp[i-1][j]) % MOD;
            }
            if (s[j] == '1') {
                dp[i][j] = (dp[i][j] + dp[i-1][j-1]) % MOD;
            }
        }
    }

    // Print the result
    std::cout << dp[N][M] << std::endl;

    return 0;
}